src/overlay/SurveyDataManager.cpp

```diff
@@ -0,0 +1,517 @@
+{
+    ZoneScoped;
+
+    if (phase() == SurveyPhase::REPORTING && mNonce == request.nonce &&
+        mSurveyor == request.request.surveyorPeerID)
+    {
+        if (!mFinalNodeData.has_value())
+        {
+            emitInconsistencyError("getSurveyData");
+            return false;
+        }
+
+        response.nodeData = mFinalNodeData.value();
+        response.inboundPeers = fillTimeSlicedPeerDataList(
+            mFinalInboundPeerData,
+            static_cast<size_t>(request.inboundPeersIndex));
+        response.outboundPeers = fillTimeSlicedPeerDataList(
+            mFinalOutboundPeerData,
+            static_cast<size_t>(request.outboundPeersIndex));
+        return true;
+    }
+    return false;
+}
+
+bool
+SurveyDataManager::surveyIsActive()
+{
+    return phase() != SurveyPhase::INACTIVE;
+}
+
+#ifdef BUILD_TESTS
+void
+SurveyDataManager::setPhaseMaxDurationsForTesting(
+    std::chrono::minutes maxPhaseDuration)
+{
+    mMaxPhaseDurationForTesting = maxPhaseDuration;
+}
+#endif
+
+SurveyPhase
+SurveyDataManager::phase()
+{
+    if (mCollectStartTime.has_value() && !mCollectEndTime.has_value())
+    {
+        if (mGetNow() >
+            mCollectStartTime.value() + getCollectingPhaseMaxDuration())
+        {
+            CLOG_INFO(
+                Overlay,
+                "Survey collecting phase has expired. Resetting survey data.");
+            reset();
+            return SurveyPhase::INACTIVE;
+        }
+        return SurveyPhase::COLLECTING;
+    }
+    if (mCollectStartTime.has_value() && mCollectEndTime.has_value())
+    {
+        if (mGetNow() >
+            mCollectEndTime.value() + getReportingPhaseMaxDuration())
+        {
+            CLOG_INFO(
+                Overlay,
+                "Survey reporting phase has expired. Resetting survey data.");
+            reset();
+            return SurveyPhase::INACTIVE;
+        }
+        return SurveyPhase::REPORTING;
+    }
+    releaseAssert(!mCollectStartTime.has_value());
+    releaseAssert(!mCollectEndTime.has_value());
+    return SurveyPhase::INACTIVE;
+}
+
+void
+SurveyDataManager::reset()
+{
+    mCollectStartTime.reset();
+    mCollectEndTime.reset();
+    mNonce.reset();
+    mSurveyor.reset();
+    mCollectingNodeData.reset();
+    mCollectingInboundPeerData.clear();
+    mCollectingOutboundPeerData.clear();
+    mFinalNodeData.reset();
+    mFinalInboundPeerData.clear();
+    mFinalOutboundPeerData.clear();
+}
+
+void
+SurveyDataManager::emitInconsistencyError(std::string const& where)
+{
+#ifdef BUILD_TESTS
+    // Throw an exception when testing to make the error more visible
+    throw std::runtime_error("Encountered inconsistent survey data while "
+                             "executing `" +
+                             where + "`.");
+#endif
+    CLOG_ERROR(Overlay, "Encountered inconsistent survey data while executing "
+                        "`{}`. Resetting survey state.");
```

Comment:
Looks like this log is missing a parameter?

Reply:
Good catch. Fixed in f92bcdfb1e7e48082942057b07203183e285f81b